### PR TITLE
generic API responses

### PR DESCRIPTION
**Pull request by bogdan-rosianu**
Changed the API response structure for all existing routes to include 3 fields:

- Data (which will contain the desired result if the request was successful)
- Error (which will contain the error message or nothing otherwise)
- Code (which will be both human and machine-readable string and will tell the code of the response - example: `bad_request`)

### CODE REVIEW

File: api/address/routes.go [hunk 1]
```diff
@@ -39,40 +40,88 @@ func Routes(router *wrapper.RouterWrapper) {
 func GetAccount(c *gin.Context) {
 	ef, ok := c.MustGet("elrondFacade").(FacadeHandler)
 	if !ok {
-		c.JSON(http.StatusInternalServerError, gin.H{"error": errors.ErrInvalidAppContext.Error()})
+		c.JSON(
+			http.StatusInternalServerError,
+			core.GenericAPIResponse{
```
Comment (iulianpascalau) [inline api/address/routes.go:45]: Why not move inside api package? Do we need this elsewhere? Core will get full of unrelated stuff.

File: api/address/routes.go [hunk 2]
```diff
@@ -39,40 +40,88 @@ func Routes(router *wrapper.RouterWrapper) {
...
 func GetBalance(c *gin.Context) {
 	ef, ok := c.MustGet("elrondFacade").(FacadeHandler)
 	if !ok {
-		c.JSON(http.StatusInternalServerError, gin.H{"error": errors.ErrInvalidAppContext.Error()})
+		c.JSON(
+			http.StatusInternalServerError,
+			core.GenericAPIResponse{
+				Data:  nil,
+				Error: errors.ErrInvalidAppContext.Error(),
+				Code:  string(core.ReturnCodeInternalError),
```
Comment (iulianpascalau) [inline api/address/routes.go:87]: return codes also should be inside api package. They are "api" related

File: api/node/routes.go [hunk 3]
```diff
@@ -67,52 +68,115 @@ func Routes(router *wrapper.RouterWrapper) {
...
+		)
 		return
 	}
 
 	details := ef.StatusMetrics().StatusMetricsMapWithoutP2P()
-	c.JSON(http.StatusOK, gin.H{"details": details})
+	c.JSON(
+		http.StatusOK,
+		core.GenericAPIResponse{
+			Data:  gin.H{"details": details},
```
Comment (iulianpascalau) [inline api/node/routes.go:149]: why wrapping it again?

File: api/node/routes.go [hunk 4]
```diff
@@ -67,52 +68,115 @@ func Routes(router *wrapper.RouterWrapper) {
...
+		)
 		return
 	}
 
 	details := ef.StatusMetrics().StatusP2pMetricsMap()
-	c.JSON(http.StatusOK, gin.H{"details": details})
+	c.JSON(
+		http.StatusOK,
+		core.GenericAPIResponse{
+			Data:  gin.H{"details": details},
```
Comment (iulianpascalau) [inline api/node/routes.go:175]: also here?

File: api/address/routes.go (hunk 1 above)
Comment (bogdan-rosianu) [reply api/address/routes.go:45]: first time I placed them in the API package but there were cyclic imports

File: api/address/routes.go (hunk 2 above)
Comment (bogdan-rosianu) [reply api/address/routes.go:87]: first time I placed them in the API package but there were cyclic imports

File: api/node/routes.go (hunk 3 above)
Comment (bogdan-rosianu) [reply api/node/routes.go:149]: in order to group them under a functionality/component. I did the same everywhere. I might rename `details` to `metrics`

File: api/node/routes.go (hunk 4 above)
Comment (bogdan-rosianu) [reply api/node/routes.go:175]: in order to group them under a functionality/component. I did the same everywhere. I might rename details to metrics

File: api/address/routes.go (hunk 2 above)
Comment (bogdan-rosianu) [reply api/address/routes.go]: created a sub-package `shared` inside the `api` package

File: api/address/routes.go (hunk 1 above)
Comment (bogdan-rosianu) [reply api/address/routes.go]: created a sub-package `shared` inside the `api` package

File: api/node/routes.go (hunk 3 above)
Comment (iulianpascalau) [reply api/node/routes.go]: no, the ideea from the meeting was that, undeer the "data" field we should have the raw object. That is what I have understood. @ccorcoveanu ?

Comment (ccorcoveanu) [reply api/node/routes.go]: A cleaner approach is to have this extra property wrap inside `data`, this way the response is verbose enough so you know what you are getting back.

File: api/address/routes.go [hunk 5]
```diff
@@ -39,40 +40,88 @@ func Routes(router *wrapper.RouterWrapper) {
 func GetAccount(c *gin.Context) {
 	ef, ok := c.MustGet("elrondFacade").(FacadeHandler)
 	if !ok {
-		c.JSON(http.StatusInternalServerError, gin.H{"error": errors.ErrInvalidAppContext.Error()})
+		c.JSON(
+			http.StatusInternalServerError,
+			shared.GenericAPIResponse{
+				Data:  nil,
+				Error: errors.ErrInvalidAppContext.Error(),
+				Code:  string(shared.ReturnCodeInternalError),
```
[the remark this answers is not in the document]
Comment (ccorcoveanu) [reply api/address/routes.go:48]: Also, return code suggests integer, so I would use integers for this. If we have the `ReturnCode` type, it could implement `Stringer` interface. We can have a method: `FormatError(code ReturnCode, err error)` that returns `code + ":" + err.Error()`, and this can be our Error for `GenericAPIResponse`.

[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply api/address/routes.go:48]: If I would set it to `ReturnCode` type, why would this help us? You can see the return code of each API call without the need to be included in the response payload. The idea of return code is that we can define custom return codes that can be customized in the future. For now, I only 'copied' the 200, 400 and 500 return codes. 
@andreibancioiu can you please provide an input?

[the remark this answers is not in the document]
Comment (ccorcoveanu) [reply api/address/routes.go:48]: I was saying to use the ReturnCode type directly in the structure so you don’t have to do string(ReturnCode)

[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply api/address/routes.go]: Done. also, changed return codes messages to snake case

File: api/node/routes.go [hunk 6]
```diff
@@ -279,10 +272,10 @@ func PeerInfo(c *gin.Context) {
 	if !ok {
 		c.JSON(
 			http.StatusInternalServerError,
```
Comment (iulianpascalau) [inline api/node/routes.go:274]: Maybe remove the struct from routes_test.go? The one with the TODO, also the TODO on L352 should be fixed. Thanks.

Comment (bogdan-rosianu) [reply api/node/routes.go:274]: done
